src/org/labkey/test/tests/SampleSetRemoteAPITest.java

```diff
@@ -181,6 +196,328 @@ public void sampleSetWithMissingValueField() throws IOException, CommandExceptio
+
+        TestDataGenerator dgen1 = new TestDataGenerator(assayLookup)
+                .withColumnSet(resultsFieldset)
+                .addCustomRow(Map.of("ParticipantID", "Jeff", "Date", "11/11/2018", "SampleName", "Green", "SampleVolume", 12.5))
+                .addCustomRow(Map.of("ParticipantID", "Jim", "Date", "11/12/2018", "SampleName", "Red", "SampleVolume", 14.5))
+                .addCustomRow(Map.of("ParticipantID", "Billy", "Date", "11/13/2018", "SampleName", "Yellow", "SampleVolume", 17.5))
+                .addCustomRow(Map.of("ParticipantID", "Michael", "Date", "11/14/2018", "SampleName", "Orange", "SampleVolume", 11.5));
+        String pasteData1 = dgen1.writeTsvContents();
+
+        TestDataGenerator dgen2 = new TestDataGenerator(assayLookup)
+                .withColumnSet(resultsFieldset)
+                .addCustomRow(Map.of("ParticipantID", "Harry", "Date", "10/11/2018", "SampleName", "Green", "SampleVolume", 12.5))
+                .addCustomRow(Map.of("ParticipantID", "William", "Date", "10/12/2018", "SampleName", "Red", "SampleVolume", 14.5))
+                .addCustomRow(Map.of("ParticipantID", "Jenny", "Date", "10/13/2018", "SampleName", "Yellow", "SampleVolume", 17.5))
+                .addCustomRow(Map.of("ParticipantID", "Hermione", "Date", "10/14/2018", "SampleName", "Orange", "SampleVolume", 11.5));
+        String pasteData2 = dgen2.writeTsvContents();
+
+        TestDataGenerator dgen3 = new TestDataGenerator(assayLookup)
+                .withColumnSet(resultsFieldset)
+                .addCustomRow(Map.of("ParticipantID", "George", "Date", "10/11/2018", "SampleName", "Green", "SampleVolume", 12.5))
+                .addCustomRow(Map.of("ParticipantID", "Arthur", "Date", "10/12/2018", "SampleName", "Red", "SampleVolume", 14.5))
+                .addCustomRow(Map.of("ParticipantID", "Colin", "Date", "10/13/2018", "SampleName", "Yellow", "SampleVolume", 17.5))
+                .addCustomRow(Map.of("ParticipantID", "Ronald", "Date", "10/14/2018", "SampleName", "Orange", "SampleVolume", 11.5));
+        String pasteData3 = dgen3.writeTsvContents();
+
+        waitAndClickAndWait(Locator.linkWithText(assayName));
+        AssayRunsPage assayRunsPage = new AssayRunsPage(getDriver());
+        DataRegionTable runsTable = assayRunsPage.getTable();
+        runsTable.clickHeaderButton("Import Data");
+        clickButton("Next");
+
+        // insert 3 runs
+        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData1);
+        clickButton("Save and Import Another Run");
+        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData2);
+        clickButton("Save and Import Another Run");
+        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData3);
+        clickButton("Save and Finish");
+
+        return Arrays.asList(dgen1, dgen2, dgen3);
+    }
+
```

Comment:
Would there be any value in making a more generic insertAssayData function? It looks like this is for tests that will always use the same data, which is useful, for future tests it might be useful to make a simpler function that allows for different data/fields:

    private void _insertAssayData(String assayName, List<TestDataGenerator> listDataGen)
    {
        waitAndClickAndWait(Locator.linkWithText(assayName));
        AssayRunsPage assayRunsPage = new AssayRunsPage(getDriver());
        DataRegionTable runsTable = assayRunsPage.getTable();
        runsTable.clickHeaderButton("Import Data");
        clickButton("Next");

        int limit = listDataGen.size();
        int imported = 0;
        for(TestDataGenerator dataGen : listDataGen)
        {
            new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", dataGen.writeTsvContents());
            imported++;
            
            if(imported < limit)
                clickButton("Save and Import Another Run");
            else
                clickButton("Save and Finish");             
        }
        
    }

This might make it more flexible for future tests that could have different columns/column-types.

Reply:
That's actually a good idea.  I like it

Reply:
Also, `AssayImportPage` has a method to save and import another run that you could chain together instead of finding the button manually. It even returns a new `AssayImportPage` for you